### PR TITLE
feat: start implementing replication - Add DOVECOT_ITERATE_ATTRS/FILTER ENVs

### PR DESCRIPTION
**Pull request by williamdes**
# Description

I added two ENVs needed to list users: https://github.com/docker-mailserver/docker-mailserver/issues/2048#issuecomment-865180458

Ref #2048

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by casperklein** [comment]
I just did a quick look and wondered, if the two newly added ENVs don't need to be added to `_setup_ldap()` @ [ldap.sh](https://github.com/docker-mailserver/docker-mailserver/blob/18f8d2573bcd8241fb77c7b1431af300c2216e4f/target/scripts/startup/setup.d/ldap.sh#L3)

Does changing those ENVs actually works?

**Comment by williamdes** [comment]
> I just did a quick look and wondered, if the two newly added ENVs don't need to be added to `_setup_ldap()` @ [ldap.sh](https://github.com/docker-mailserver/docker-mailserver/blob/18f8d2573bcd8241fb77c7b1431af300c2216e4f/target/scripts/startup/setup.d/ldap.sh#L3)
> 
> Does changing those ENVs actually works?

Changing the ENVs does work because the tests say so and my production ENV too since some years now ;p
That said, I am wondering if this is only used by doveadm to list users
or to list and manage them
or to also provide a list of users to replicate
I did not check this since I do not want to break my current production replication.

But most probably it's needed to list users to replicate.
Ref: https://github.com/docker-mailserver/docker-mailserver/issues/2048#issuecomment-865180458

**Comment by casperklein** [comment]
You're right. I found the snipped that handles the ENV change: https://github.com/docker-mailserver/docker-mailserver/blob/18f8d2573bcd8241fb77c7b1431af300c2216e4f/target/scripts/startup/setup.d/ldap.sh#L56

**Review by georglauterbach — DISMISSED** [review]
LGTM 👍🏼

**Review by polarathene — CHANGES_REQUESTED** [review]
This looks fine, except for the test which is not acceptable. Not your fault as I don't expect you to have known why :sweat_smile: 

---

Most tests have been ported over to our newer conventions which makes it easier to write / maintain tests. LDAP is one of the few that I didn't get around to tackling with my available volunteer time for DMS.

Here you've brought in the same `common` helper script but from the revised version. You then leverage `X_EXTRA_ARGS` in a way that it's not meant to be used :sweat_smile: (_you probably wanted to add the `setup` helper and call [our dedicated method for that](https://github.com/docker-mailserver/docker-mailserver/blob/7d5c2736cec0cc6db198da57e94a8ca589570d9f/test/helper/setup.bash#L151-L172) instead_)

Here's an example of the newer version of tests for `setup_file()`:

https://github.com/docker-mailserver/docker-mailserver/blob/7d5c2736cec0cc6db198da57e94a8ca589570d9f/test/tests/serial/mail_with_imap.bats#L1-L21

LDAP has a little bit extra going on with a custom network and LDAP container also being setup.

I can also see that your new test cases are calling `setup_file()` directly with additional ENV. `setup_file()` is only meant to be run once for the whole file, test cases aren't meant to call it again.

`setup_file()` will do any common setup that's relevant for all test cases before they're run, when you need to run containers with different ENV, you can likewise setup the container in test cases instead of in `setup_file()`:

https://github.com/docker-mailserver/docker-mailserver/blob/d3755aeddc316139675e4741d304f66e38029151/test/tests/parallel/set3/container_configuration/process_check_restart.bats#L133-L143

Then the teardown will happen automatically if our test file has:

https://github.com/docker-mailserver/docker-mailserver/blob/d3755aeddc316139675e4741d304f66e38029151/test/tests/parallel/set3/container_configuration/process_check_restart.bats#L9

And you've set the container name as shown above.

**Comment by williamdes** [comment]
Thanks for the feedback @polarathene

Since I mostly did follow the actual testing style (of the file) could we merge this PR and cleanup this in another one?
I would like to avoid having open contributions to look after

**Comment by georglauterbach** [comment]
Fine we me, just sayin :)

**Comment by polarathene** [comment]
> Since I mostly did follow the actual testing style (of the file) could we merge this PR and cleanup this in another one?

Are you actually going to follow-up with it?

It's not about testing style. I pointed out issues with how you modified the test. You are using the test suite methods itself incorrectly, not just our helpers / conventions.

I don't think we can trust the tests are reliable with the current concerns raised.

---

You can wait until end of the month, I may have time to address this myself. Otherwise if someone does merge this PR ignoring my concerns:
- Include a disclaimer comment at the top of the test file referencing my earlier comment about test quality concerns.
- Potentially revert all other changes to the test beyond the new test cases and comment them out, with comment referencing the same PR comment about the quality of the approach not being acceptable.

**Comment by williamdes** [comment]
Oh gosh I forgot about this
I would really appreciate if someone can merge this and patch the test part before me

**Comment by georglauterbach** [comment]
> Oh gosh I forgot about this I would really appreciate if someone can merge this and patch the test part before me

This is probably not going to happen..

**Comment by williamdes** [comment]
> > Oh gosh I forgot about this I would really appreciate if someone can merge this and patch the test part before me
> 
> This is probably not going to happen due to a simple lack of time..

Nevermind, still on my todo list

**Comment by polarathene** [comment]
I've been working on the LDAP test today, it'll be refactored and should be reviewed / merged this week sometime. Your PR here can then adapt to the new test.

**Comment by williamdes** [comment]
> I've been working on the LDAP test today, it'll be refactored and should be reviewed / merged this week sometime. Your PR here can then adapt to the new test.

Thank you so much

**Comment by polarathene** [comment]
@williamdes while we wait on https://github.com/docker-mailserver/docker-mailserver/pull/3483 to get merged, are you familiar with any other openldap docker images that you can easily automate setup with?

The one we're using in our tests is quite outdated (version pinned), and upstream is no longer maintained for 2 years.

I have come across a bitnami / vmware image, but it doesn't support `.schema` files (`postfix-book.schema`). I figured out how to convert that to a `.ldif` via `slaptest` + `slapcat` I think, but still couldn't quite get the image happy as it was failing with "Invalid DN syntax" when parsing the `.ldif` files we have for creating the test users.

**Comment by williamdes** [comment]
> @williamdes while we wait on #3483 to get merged, are you familiar with any other openldap docker images that you can easily automate setup with?
> 
> The one we're using in our tests is quite outdated (version pinned), and upstream is no longer maintained for 2 years.
> 
> I have come across a bitnami / vmware image, but it doesn't support `.schema` files (`postfix-book.schema`). I figured out how to convert that to a `.ldif` via `slaptest` + `slapcat` I think, but still couldn't quite get the image happy as it was failing with "Invalid DN syntax" when parsing the `.ldif` files we have for creating the test users.

Yeah of course, I built one for the same reasons
https://github.com/sudo-bot/docker-openldap/
With the old one it should be plug and play
See it implemented with DMS: https://github.com/datacenters-network/mails/blob/a28cc971351bad2ee98fe63e04cb2879aeeb8877/docker-compose.yml#L173 
Since years now! 

Please file issues with anything you want to have for it 
Yes I need to add documentation ^^

**Comment by polarathene** [comment]
The LDAP test refactor PR is merged, I'll revise the portion of your test changes in this PR by the weekend 👍 

---

I managed to resolve the issue I was facing today. I gave your image a look but didn't feel it was a better choice for DMS maintainers to rely on. Detailed notes below.

> With the old one it should be plug and play

Doesn't seem that way. I looked through the reference links you provided but stopped by this point:

```bash
docker run --rm --name ldap-test \
  --env LDAP_ADMIN_PASSWORD=admin --env LDAP_CONFIG_PASSWORD=admin --env LDAP_MONITOR_PASSWORD=admin \
  --env LDAP_BASE_DN='dc=example,dc=test' \
  --env LDAP_AUTH_BASE_DN='ou=people,dc=example,dc=test' \
  --volume '/tmp/ldifs/:/ldifs/:ro' \
  --hostname 'ldap.example.test' \
  botsudo/docker-openldap
```

It wanted me to configure more ENV and refused to exit with `ctrl+c`. Also looking over your reference links, it was unclear how the custom data would be populated by the image beyond the seeding example where you have extra scripts to handle it.

The image itself seems to have a lot more going on package / config wise, I would have liked to get the [`bitnami/openldap` image](https://github.com/bitnami/containers/blob/main/bitnami/openldap/README.md) working, but it kept exiting when supplying the `.ldif` account files:

### Alternative (_most popular openldap image actively maintained_): `bitnami/openldap`

```bash
 docker run --rm --name ldap-test \
  --env LDAP_ROOT='dc=example,dc=test' \
  --env LDAP_PORT_NUMBER=389 \
  --env BITNAMI_DEBUG=true \
  --volume '/tmp//ldifs/:/ldifs/' \
  --volume '/tmp/schema/postfix-schema.ldif:/schema/custom.ldif:ro' \
  --hostname 'ldap.example.test' \
  bitnami/openldap:latest
```

That image lets you supply a directory with `.ldif` files to seed with, but the image doesn't support `.schema` files apparently. The advice I had come across was to convert `.schema` into `.ldif`, so I have done that for the `postfix-book.schema` (_and volume mounted above per the image README_):

```bash
# Read a conf file that imports the schema and any dependencies needed, then export:
mkdir /tmp/ldif-output
slaptest -f /postfix-schema.conf -F /tmp/ldif-output

# Reads converted schema output and filters out unnecessary properties, saving to ldif file:
slapcat -o ldif_wrap=no -n 0 -F /tmp/ldif -H "ldap:///???(cn={4}postfix-book)" \
  | grep -v ^structuralObjectClass: \
  | grep -v ^entryUUID: \
  | grep -v ^creatorsName: \
  | grep -v ^createTimestamp: \
  | grep -v ^entryCSN: \
  | grep -v ^modifiersName: \
  | grep -v ^modifyTimestamp: \
  | sed -e "s/{4}postfix-book/postfix-book/" > /tmp/postfix-schema.ldif
```

**`/postfix-schema.conf`:**
```
include /opt/bitnami/openldap/etc/schema/core.schema
include /opt/bitnami/openldap/etc/schema/cosine.schema
include /opt/bitnami/openldap/etc/schema/inetorgperson.schema
include /opt/bitnami/openldap/etc/schema/nis.schema
include /postfix-book.schema
```

**`/tmp/postfix-schema.ldif`:**
```
dn: cn=postfix-book,cn=schema,cn=config
objectClass: olcSchemaConfig
cn: postfix-book
olcAttributeTypes: {0}( 1.3.6.1.4.1.29426.1.10.1 NAME 'mailHomeDirectory' DESC 'The absolute path to the mail user home directory' EQUALITY caseExactIA5Match SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 SINGLE-VALUE )
olcAttributeTypes: {1}( 1.3.6.1.4.1.29426.1.10.2 NAME 'mailAlias' DESC 'RFC822 Mailbox - mail alias' EQUALITY caseIgnoreIA5Match SUBSTR caseIgnoreIA5SubstringsMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{256} )
olcAttributeTypes: {2}( 1.3.6.1.4.1.29426.1.10.3 NAME 'mailUidNumber' DESC 'UID required to access the mailbox' EQUALITY integerMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE )
olcAttributeTypes: {3}( 1.3.6.1.4.1.29426.1.10.4 NAME 'mailGidNumber' DESC 'GID required to access the mailbox' EQUALITY integerMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE )
olcAttributeTypes: {4}( 1.3.6.1.4.1.29426.1.10.5 NAME 'mailEnabled' DESC 'TRUE to enable, FALSE to disable account' EQUALITY booleanMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.7 SINGLE-VALUE )
olcAttributeTypes: {5}( 1.3.6.1.4.1.29426.1.10.6 NAME 'mailGroupMember' DESC 'Name of a mail distribution list' EQUALITY caseExactIA5Match SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 )
olcAttributeTypes: {6}( 1.3.6.1.4.1.29426.1.10.7 NAME 'mailQuota' DESC 'Mail quota limit in kilobytes' EQUALITY caseExactIA5Match SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 )
olcAttributeTypes: {7}( 1.3.6.1.4.1.29426.1.10.8 NAME 'mailStorageDirectory' DESC 'The absolute path to the mail users mailbox' EQUALITY caseExactIA5Match SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 SINGLE-VALUE )
olcObjectClasses: {0}( 1.3.6.1.4.1.29426.1.2.2.1 NAME 'PostfixBookMailAccount' DESC 'Mail account used in Postfix Book' SUP top AUXILIARY MUST mail MAY ( mailHomeDirectory $ mailAlias $ mailGroupMember $ mailUidNumber $ mailGidNumber $ mailEnabled $ mailQuota $ mailStorageDirectory ) )
olcObjectClasses: {1}( 1.3.6.1.4.1.29426.1.2.2.2 NAME 'PostfixBookMailForward' DESC 'Mail forward used in Postfix Book' SUP top AUXILIARY MUST ( mail $ mailAlias ) )
```

Even if that's one of the first items to be parsed lumped in with the `/ldifs` mount dir, it still has problems 🤷‍♂️ Yet if I shell in to the container instead with those files mounted at a different location and manually add them with `ldapadd` no problem. I'd have to troubleshoot further, but supposedly just mounting the `/ldifs` dir alters the initial config setup.

One issue on their repo suggested using their supported `/docker-entrypoint-initdb.d/` to add in a custom shell script that would do the necessary `ldapadd`.

My previous comment suggests I may have got it working at some point, but not quite with the "Invalid DN syntax" issue.

### Alternative: `bitnami/openldap` image variant 

Based off `bitnami/openldap`: https://github.com/clayrisser/docker-openldap

This image carries the `postfix-book.schema` among other improvements, and provides the best experience so far, although it's a bit to new but for using in tests should be fine 👍 

```bash
 docker run --rm --name ldap-test \
  --env LDAP_ROOT='dc=example,dc=test' \
  --env LDAP_PORT_NUMBER=389 \
  --env BITNAMI_DEBUG=true \
  --volume '/tmp//ldifs/:/ldifs/' \
  --volume '/tmp/migrations:/migrations:ro' \
  --hostname 'ldap.example.test' \
  registry.gitlab.com/bitspur/rock8s/docker-openldap
```

The only issue I ran into was getting `testsaslauthd` to be successful. Got familiar with LDAP a bit more and some commands, and eventually figured out the issue was the default ACL config was causing a [misleading "Invalid credentials (49)" error](https://github.com/clayrisser/docker-openldap/issues/1#issuecomment-1681695098).

The separate `migrations/` volume provides an `.ldif` that corrects that based on our current `osixia/openldap` image used and related openldap docs I came across.

**Comment by williamdes** [comment]
Can you open an issue on my image?

If I understand you are missing loading ldiff files at startup?

You "should not need" to load more of them since they are pre loaded by myself, the postfix schema and all.
No more need for a custom image
That said I did some tweaks to have saslauthd inside it, maybe I should remove this part in a variant of the image

**Comment by williamdes** [comment]
> The LDAP test refactor PR is merged, I'll revise the portion of your test changes in this PR by the weekend 👍

On it

File: test/tests/serial/mail_with_ldap.bats
Comment (williamdes) [inline test/tests/serial/mail_with_ldap.bats:174]: for reasons that I ignore, this does not seem to give the options to the container creation process

Any help would be very welcome 🙏🏻

Comment (georglauterbach) [reply test/tests/serial/mail_with_ldap.bats:174]: You will need to provide the array `CUSTOM_SETUP_ARGUMENTS` on container startup to the function that creates the container. Functions like `_run_in_container` do not take it into account.

If you want to have it added to the ENV of the container, you'll need to do so at the top of the file where the container is created I guess (or wherever it is created). Probably

https://github.com/docker-mailserver/docker-mailserver/pull/3426/files#diff-664e916443d73f328b2c0c61c3a1aca1de5b968cfb03316012765971cbb5980bR79

and

https://github.com/docker-mailserver/docker-mailserver/pull/3426/files#diff-664e916443d73f328b2c0c61c3a1aca1de5b968cfb03316012765971cbb5980bR93

**Comment by polarathene** [comment]
> On it

Yeah sorry, I've had some other events need my attention and within DMS I've been trying to queue some additional LDAP related changes before adjusting the test a bit more before meeting your requirements needed here. Almost there 😅

File: test/tests/serial/mail_with_ldap.bats
Comment (polarathene) [inline test/tests/serial/mail_with_ldap.bats:195]: Here's an example of what this would look like from the advice given in my review feedback:

```bash
@test "dovecot: (ENV DOVECOT_ITERATE_FILTER=(objectClass=PostfixBookMailAccount)) ldap email listing works" {
  # Container setup:
  export CONTAINER_NAME='dms-test_ldap_dovecot-iter3'
  local CONTAINER_ARGS_ENV_CUSTOM=(
    "${CONFIG_CONTAINER_BASE[@]}"
    --env DOVECOT_ITERATE_FILTER='(objectClass=PostfixBookMailAccount)'
  )
  _init_with_defaults
  _common_container_setup 'CONTAINER_ARGS_ENV_CUSTOM'

  # Actual test-case:
  _run_in_container doveadm user '*'
  assert_success
  assert_line --index 0 'some.user@localhost.localdomain'
  assert_line --index 1 'some.other.user@localhost.otherdomain'
  assert_line --index 2 'some.user.email@localhost.localdomain'
  _should_output_number_of_lines 3

  # Cleanup, removes test-case container:
  _default_teardown
}
```

The cleanup is just force removing the container:

https://github.com/docker-mailserver/docker-mailserver/blob/e9f04cf8a7d31efbc8ae9a4a35603409f637f3b4/test/helper/setup.bash#L198-L204

Setup is like we do in `setup()`, except we need to export the custom args as a common base var to expand / import here. It's important to give each container a unique name to identify it, I chose to use the name variant `dovecot-iter3` as it's the 3rd test-case container for testing against that ENV (_or 2nd if we don't create a separate container for the default test_).

Setting the name will ensure the helper methods used will know which container to use implicitly.

**Review by polarathene — CHANGES_REQUESTED** [review]
You can probably use this review feedback to get the tests working/passing, but like I've said I'm combing through LDAP support and history within DMS and may introduce some breaking changes before returning to this PR.

I don't feel there is a rush to merge this, unless you have a follow-up you want to push through for v13 as well? (_if so, take my feedback and update the LDAP tests for me and we'll look into merging it sooner_)

**Comment by polarathene** [comment]
Once [this PR is merged](https://github.com/docker-mailserver/docker-mailserver/pull/3514), adding your LDAP test cases should be pretty straight-forward 😁 

- Add `CONTAINER0_NAME=` (where `0` is incremented appropriately) with a value matching the convention shown at the top of the test file. See the referenced PR for usage with the extra container added for a test-case.
- Add container config/setup at end of `setup()`, just before the final `export` (_used to set `CONTAINER_NAME` to the main DMS container_).
- Add your test-cases with a small change like the modified one shows (_should only require adding the extra `export CONTAINER_NAME=` line_).

**Comment by polarathene** [comment]
I am [considering a different approach](https://github.com/docker-mailserver/docker-mailserver/issues/3517#issuecomment-1702053878) that should meet your needs for these two settings, and any other users future needs for missing settings that these configs are meant to support. Our current approach is a bit awkward and could be improved both for the user and maintainers.

**Comment by williamdes** [comment]
Closing this one. My brain forgot what was left todo.

Feel free to pick stuff out if needed

**Comment by williamdes** [comment]
Thank you for your work on LDAP support @polarathene !

Please do mention me directly on PRs so I can review them if you want.

Having an easier way to define ENVs to create the dovecot config keys 

> @williamdes how important is the ENV support to set these two settings vs including a file with them? I understand your request for this was to https://github.com/docker-mailserver/docker-mailserver/issues/2048#issuecomment-1568514738.

The only goal was to remove more lines from my patch file :)

**Comment by georglauterbach** [comment]
Thank you for closing this as it has become stale! This saves me the maintenance of stale PRs :)

**Comment by polarathene** [comment]
> Please do mention me directly on PRs so I can review them if you want.

https://github.com/docker-mailserver/docker-mailserver/pull/3524 is the improved approach to what this PR was tackling. It enables support more broadly. Feel free to review it or suggest feedback for some concerns I've noted.

It's only really blocked by my rewrite of the docs for the feature which I've kept on-hold due to very low time right now. Hoping to have a week or so for dedicating to DMS later this month.

**Pull request closed without merging**
